pkg/blockbuilder/blockbuilder_test.go

```diff
@@ -49,25 +55,144 @@ func TestBlockBuilder_WipeOutDataDirOnStart(t *testing.T) {
+
+	reg := prometheus.NewPedanticRegistry()
+	bb, err := New(cfg, test.NewTestingLogger(t), reg, overrides)
+	require.NoError(t, err)
+
+	require.NoError(t, services.StartAndAwaitRunning(ctx, bb))
+	t.Cleanup(func() {
+		require.NoError(t, services.StopAndAwaitTerminated(ctx, bb))
+	})
+
+	require.Eventually(t, func() bool {
+		return assert.NoError(t, promtest.GatherAndCompare(reg, strings.NewReader(`
+			# HELP cortex_blockbuilder_consumer_lag_records The per-topic-partition number of records, instance needs to work through each cycle.
+			# TYPE cortex_blockbuilder_consumer_lag_records gauge
+			cortex_blockbuilder_consumer_lag_records{partition="0",topic="test"} 1
+			cortex_blockbuilder_consumer_lag_records{partition="1",topic="test"} 0
+		`), "cortex_blockbuilder_consumer_lag_records"))
+	}, 5*time.Second, 100*time.Millisecond)
+}
+
+// Testing block builder starting up with an existing kafka commit.
+func TestBlockBuilder_StartWithExistingCommit(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithCancelCause(context.Background())
+	t.Cleanup(func() { cancel(errors.New("test done")) })
+
+	_, kafkaAddr := testkafka.CreateClusterWithoutCustomConsumerGroupsSupport(t, numPartitions, testTopic)
+
+	kafkaClient := mustKafkaClient(t, kafkaAddr)
+	kafkaClient.AddConsumeTopics(testTopic)
+
+	cfg, overrides := blockBuilderConfig(t, kafkaAddr)
+
+	// Producing some records
+	cycleEndStartup := cycleEndAtStartup(time.Now(), cfg.ConsumeInterval, cfg.ConsumeIntervalBuffer)
+
+	var producedSamples []mimirpb.Sample
+	kafkaRecTime := cycleEndStartup.Truncate(cfg.ConsumeInterval).Add(-7 * time.Hour).Add(29 * time.Minute)
+	for kafkaRecTime.Before(cycleEndStartup) {
+		samples := floatSample(kafkaRecTime.Add(-time.Minute).UnixMilli(), 0)
+		val := createWriteRequest(t, "", samples, nil)
+
+		produceRecords(ctx, t, kafkaClient, kafkaRecTime, "1", testTopic, 0, val)
+		producedSamples = append(producedSamples, samples...)
+
+		kafkaRecTime = kafkaRecTime.Add(cfg.ConsumeInterval / 2)
+	}
+	require.NotEmpty(t, producedSamples)
+
+	// Fetch all the records that were produced to choose one to commit.
+	var recs []*kgo.Record
+	for len(recs) < len(producedSamples) {
+		fetches := kafkaClient.PollFetches(ctx)
+		require.NoError(t, fetches.Err())
+		recs = append(recs, fetches.Records()...)
+	}
+	require.Len(t, recs, len(producedSamples))
+
+	// Choosing the midpoint record to commit and as the last seen record as well.
+	commitRec := recs[len(recs)/2]
+	require.NotNil(t, commitRec)
+
+	lastRec := commitRec
+	blockEnd := commitRec.Timestamp.Truncate(cfg.ConsumeInterval).Add(cfg.ConsumeInterval)
+	offset := kadm.Offset{
+		Topic:       commitRec.Topic,
+		Partition:   commitRec.Partition,
+		At:          commitRec.Offset + 1,
+		LeaderEpoch: -1, // not a group consumer
+		Metadata:    marshallCommitMeta(commitRec.Timestamp.UnixMilli(), lastRec.Timestamp.UnixMilli(), blockEnd.UnixMilli()),
+	}
+	commitOffset(ctx, t, kafkaClient, testGroup, offset)
```

Comment:
did you consider just running the block builder multiple times? That's a more block-box approach and also tests that the same block builder can pick up from where it left off. Here we copy the logic for commiting and that can break later on if the BB start commiting in a different way.

i think it's mostly a matter of preference, but wanted to hear your thoughts on it